src/docker/restorableappengine.cc

```diff
@@ -443,9 +443,9 @@ void RestorableAppEngine::pullApp(const Uri& uri, const boost::filesystem::path&
   registry_client_->downloadBlob(archive_uri, archive_full_path, manifest.archiveSize());
   Utils::writeFile(app_dir / Manifest::Filename, manifest_str);
   Utils::writeFile(app_dir / "uri", uri.registryHostname + "/" + uri.repo + "@" + uri.digest());
-  // extract docker-compose.yml, temporal hack, we don't need to extract it
-  exec(boost::format{"tar -xzf %s %s"} % archive_full_path % ComposeFile, "no compose file found in archive",
-       boost::process::start_dir = app_dir);
+  // Extract docker-compose.yml and safely persist it so the follow-up functionality doesn't need to do it again.
+  const auto compose{extractComposeFile(archive_full_path)};
+  Utils::writeFile(app_dir / ComposeFile, compose);
```

Comment:
orthogonal - but i noticed a bug in `writeFile` - https://github.com/foundriesio/aktualizr/blob/2023.06%2Bfio/src/libaktualizr/utilities/utils.cc#L337-L339 there needs to be a call to `sync` between the file close (or if the ostream could call sync itself) and the rename or rename could wind up pointing at an inode with no data

Reply:
I think file.close() includes sync.

Reply:
That said, I'm not sure how critical it is for us.

The `fsync` call can still result into another _buffering layer_ inside a disk cache; but POSIX compliant disk manufacturers must flush that cache in case of fsync.

At the very least `fsync` call would warrant that all readers will see the "updated" version of the file.

Reply:
the fsync is required for power safe operations. Its a very unlikely thing, but its the only way. `rename` is an atomic call - so it will persist across a weird power fail. but if you haven't done the fsync - it can point at garbage.

Reply:
>The file.close includes a sync, but not fsync.

That's correct. I just don't know how really is important to do it (fsync), and what would be performance cost.

Reply:
Also, it's not really critical as the aklite will just re-extract a compose file from an App archive.

Reply:
One scenario I was interested in is this:
1. Application opens a file, writes to it and calls close (which calls sync).
2. Application opens a file and reads from it before the OS actually syncs the file to disk.

I'm not sure if OS gives any warranty that at step 2 application will see the writes from step 1.
Normally, it should be the case; but here we're talking about esoteric OSes on low-profile devices.
(The file rename is irrelevant here, as it will point to the same node; I'm only talking about the write to the node itself).

The fsync would give such warranties; the close - I'm not sure.

Reply:
you don't need the fsync for that scenario - the OS will handle that. `fsync` handles one specific thing:
1) write file (but content is still not flushed to disk)
2) rename file (this is okay - the os has these buffers)
3) power fails before buffers are flushed to disk

When the system starts back up step 2 will be done (there will be a file with that name) but the inodes it's pointing at won't have correct data.

`writeFile` may be "safe enough" for us - depends on what's using it - but its not safe to power failures.

Reply:
Normally, sync/close guarantees that other processes read the same content that has been written before close/sync. It should be handled at a file system level/implementation.

It doesn't guarantee that the written content is persisted on disk if a power cut happens after sync/close. That's what `fsync` is needed for - it writes down the cached data at a file system level to a disk.